### DISCUSSION TITLE
`Field(init=False)` on BaseModel

### DISCUSSION THREAD

**Discussion by felix-hilden** [pydantic/pydantic#9341] [Question]
Hi, some time ago I asked for `Field(init=False)` to be implemented into Pydantic, and it's been done recently! Thank you 🙏  A short discussion was had here: pydantic/pydantic#5929.

I was finally having a go at it, and noticed that it's only valid for dataclasses ([doc](https://docs.pydantic.dev/latest/api/fields/#pydantic.fields.Field)). So I was wondering if there's a reason for that vs. having it available in BaseModel too. Or maybe I've just missed something 😅 I can't really find any hints in pydantic/pydantic#8552.

**Comment by ilanbiala** [comment]
This also seems like a miss - it would be great to have support for in `BaseModel` as well to allow for fields to be initialized in a model's `__post_init__()` method if processing is needed.

**Comment by metab0t** [comment]
It would be great to have this supported for `BaseModel` to allow users to have fields that are not initialized in constructor.

**Comment by vinicq** [comment]
The docs are accurate. `init=False` is only honored on dataclasses today. On a plain `BaseModel` it isn't enforced, and the rationale from the linked threads is that `BaseModel`'s lifecycle is "validate then construct", not "construct then post_init", so the "a field that exists but cannot be passed to `__init__`" semantics don't map onto the model the way they do for a dataclass.

Workarounds that have held up in production code.

For a field whose value should be derived after construction, `model_validator(mode='after')` is the closest analogue to a `__post_init__`:

```python
from pydantic import BaseModel, model_validator
from typing import Self

class Item(BaseModel):
    name: str
    slug: str = ""

    @model_validator(mode="after")
    def _derive_slug(self) -> Self:
        self.slug = self.name.lower().replace(" ", "-")
        return self
```

For a read-only derivation that doesn't need to be stored, `computed_field`:

```python
from pydantic import BaseModel, computed_field

class Item(BaseModel):
    name: str

    @computed_field
    @property
    def slug(self) -> str:
        return self.name.lower().replace(" ", "-")
```

If you really need a value that cannot be derived and shouldn't appear in dumps, `default_factory` with `Field(exclude=True)`:

```python
from pydantic import BaseModel, Field

class Item(BaseModel):
    name: str
    internal_id: str = Field(default_factory=_generate_id, exclude=True)
```

What you originally asked for (`init=False` on `BaseModel`) would require dataclass-style `__post_init__` machinery, which is a bigger lifecycle change than a single field flag. Whether that's worth filing as a feature request again probably depends on how often the workarounds above feel like cruft for the code you maintain.

**Comment by s-chudmunge** [comment]
`Field(init=False)` is supported on Pydantic **dataclasses** but intentionally not on `BaseModel`. The lifecycle difference is the reason.

A dataclass has a `__post_init__` phase — you can declare a field with `init=False`, skip it in `__init__`, and set it in `__post_init__`. A `BaseModel` works differently: **validation and construction happen in a single pass** inside `__pydantic_validator__.validate_python`. There is no separate post-construction phase to write a derived value into.

## What to use instead

**Derived value computable on the fly** → `computed_field`:

```python
from pydantic import BaseModel, computed_field

class Item(BaseModel):
    name: str

    @computed_field
    @property
    def slug(self) -> str:
        return self.name.lower().replace(" ", "-")
```

Appears in `model_dump()` and JSON, always in sync. No stored field needed.

---

**Derived value that must be stored** → `model_validator(mode="after")`:

```python
from pydantic import BaseModel, model_validator
from typing import Self

class Item(BaseModel):
    name: str
    slug: str = ""

    @model_validator(mode="after")
    def _derive_slug(self) -> Self:
        self.slug = self.name.lower().replace(" ", "-")
        return self
```

Runs after all fields are populated. The `slug = ""` default keeps the type annotation honest and lets you round-trip JSON that already has a `slug`.

---

**Internal state that should not appear in output** → `default_factory` + `Field(exclude=True)`:

```python
import uuid
from pydantic import BaseModel, Field

class Item(BaseModel):
    name: str
    _internal_id: str = Field(default_factory=lambda: str(uuid.uuid4()), exclude=True)
```

---

If you genuinely need the dataclass behaviour (field absent from `__init__`, set in a post-construction hook), use `pydantic.dataclasses.dataclass` instead of `BaseModel` — `Field(init=False)` is honoured there and the result is still compatible with FastAPI and Pydantic integrations.
